adapters/rest-adapter-generator/src/main/java/uk/gov/justice/services/adapters/rest/generator/JaxRsImplementationGenerator.java

```diff
@@ -0,0 +1,348 @@
+    }
+
+    /**
+     * Uses the type of {@link Action} to provide a {@link SynchronousDispatcher} or an {@link
+     * AsynchronousDispatcher} and provides the correct field definition as a {@link FieldSpec}.
+     *
+     * @param action the action to forEach
+     * @return a {@link FieldSpec} representing the field definition.
+     * @throws IllegalStateException if action type is not GET or POST
+     */
+    private FieldSpec dispatcherFieldFor(final Action action) {
+        final ActionType actionType = action.getType();
+
+        if (actionType == GET) {
+            return synchronousDispatcherField();
+        } else if (actionType == POST) {
+            return asynchronousDispatcherField();
+        } else {
+            throw new IllegalStateException(format("Unsupported action type %s", actionType));
+        }
+    }
+
+    /**
+     * Generate the SynchronousDispatcher injected field.
+     *
+     * @return the {@link FieldSpec} that represents the generated field
+     */
+    private FieldSpec synchronousDispatcherField() {
+        return FieldSpec.builder(SynchronousDispatcher.class, "syncDispatcher")
+                .addAnnotation(Inject.class)
+                .build();
+    }
+
+    /**
+     * Generate the AsynchronousDispatcher injected field.
+     *
+     * @return the {@link FieldSpec} that represents the generated field
+     */
+    private FieldSpec asynchronousDispatcherField() {
+        return FieldSpec.builder(AsynchronousDispatcher.class, "asyncDispatcher")
+                .addAnnotation(Inject.class)
+                .build();
+    }
+
+    /**
+     * Add MimeType specific parameter to method.
+     *
+     * @param methodBuilder add parameter to this method builder
+     * @return the method builder
+     */
+    private MethodSpec.Builder addToMethodWithMimeType(final MethodSpec.Builder methodBuilder) {
+        return methodBuilder.addParameter(ParameterSpec
+                .builder(JsonObject.class, GENERIC_PAYLOAD_ARGUMENT_NAME)
+                .build());
+    }
+
+    /**
+     * Generate a method for each {@link Action}.
+     *
+     * @param action             the action to generate as a method
+     * @param resourceMethodName the resource method name to generate
+     * @return a {@link MethodSpec} that represents the generated method
+     */
+    private MethodSpec.Builder generateResourceMethod(final Action action,
+                                                      final String resourceMethodName) {
+        final ActionType actionType = action.getType();
+        final Map<String, QueryParameter> queryParams = action.getQueryParameters();
+        final Map<String, UriParameter> pathParams = action.getResource().getUriParameters();
+        final CodeBlock methodBody;
+
+        if (actionType == GET) {
+            methodBody = methodBody(pathParams, () -> methodBodyForGet(queryParams));
+        } else {
```

Comment:
The problem here is that the action type is already checked before this method is run.  So, if I add the check here it becomes untestable code.  I have decided to make it clearer what is occurring and I have added a method to validate the action type prior to processing the action.